loadtest/loadtest.go

```diff
@@ -0,0 +1,328 @@
+	if config.Clients.ControlPlaneHelmClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Clients.ControlPlaneHelmClient must not be empty", config)
+	}
+	if config.Clients.ControlPlaneK8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Clients.ControlPlaneK8sClient must not be empty", config)
+	}
+	if config.GuestFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GuestFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.AuthToken == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AuthToken must not be empty", config)
+	}
+	if config.ClusterID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterID must not be empty", config)
+	}
+	if config.CommonDomain == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CommonDomain must not be empty", config)
+	}
+
+	s := &LoadTest{
+		clients:        config.Clients,
+		guestFramework: config.GuestFramework,
+		logger:         config.Logger,
+
+		authToken:    config.AuthToken,
+		clusterID:    config.ClusterID,
+		commonDomain: config.CommonDomain,
+	}
+
+	return s, nil
+}
+
+func (l *LoadTest) Test(ctx context.Context) error {
+	var err error
+
+	var loadTestEndpoint string
+	{
+		loadTestEndpoint = fmt.Sprintf("loadtest-app.%s.%s", l.clusterID, l.commonDomain)
+
+		l.logger.Log("level", "debug", "message", fmt.Sprintf("loadtest-app endpoint is %#q", loadTestEndpoint))
+	}
+
+	{
+		l.logger.LogCtx(ctx, "level", "debug", "message", "enabling HPA for Nginx Ingress Controller")
+
+		err = l.enableIngressControllerHPA(ctx)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		l.logger.LogCtx(ctx, "level", "debug", "message", "enabled HPA for Nginx Ingress Controller")
+	}
+
+	{
+		l.logger.LogCtx(ctx, "level", "debug", "message", "installing loadtest app")
+
+		err = l.installTestApp(ctx, loadTestEndpoint)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		l.logger.LogCtx(ctx, "level", "debug", "message", "installed loadtest app")
+	}
+
+	{
+		l.logger.LogCtx(ctx, "level", "debug", "message", "waiting for loadtest app to be ready")
+
+		err = l.waitForLoadTestApp(ctx)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		l.logger.LogCtx(ctx, "level", "debug", "message", "loadtest app is ready")
+	}
+
+	return nil
+}
+
+// enableIngressControllerHPA enables HPA via the user configmap and updates
+// the chartconfig CR so chart-operator reconciles the config change.
+func (l *LoadTest) enableIngressControllerHPA(ctx context.Context) error {
+	var err error
+
+	l.logger.Log("level", "debug", "message", fmt.Sprintf("waiting for %#q configmap to be created", UserConfigMapName))
+
+	o := func() error {
+		lo := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("app=%s,cluster-operator.giantswarm.io/configmap-type=user", UserConfigMapName),
+		}
```

Comment:
I recently figured out that you can also do something like this. The snippet is just copied from somewhere but this may be a better approach. 

```
	labelSelector := metav1.AddLabelToSelector(&v1.LabelSelector{}, label.Cluster, key.ClusterID(&cluster))
	listOpts := metav1.ListOptions{
		LabelSelector: labelSelector.String(),
	}
```

Reply:
Thx, that's nicer. I'll update.

Reply:
I kept the original approach as I need to select on 2 labels. But this is neat and I'll keep it in mind for the future.